### ISSUE TITLE
99.9% cpu load when using the epoll reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2809]
|<img alt="mvw's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| mvw reported|
|-|-|
|Trac ID|trac#2809|
|Type|defect|
|Created|2007-08-31 15:06:09Z|

We observed a 99.9% cpu load for a twisted application using the epoll reactor 
on Linux but had no explanation where the time was wasted.

A kernel trace yielded many calls like this:

```
  epoll_wait(5, {{EPOLLERR|EPOLLHUP, {u32=191, u64=191```, 297, 379) = 1
}}}

The normal behaviour should be that sockets in this state won't be handled any longer.

A discussion on #twisted revealed that the epoll reactor developers 
had not encountered this situation 
where EPOLLERR and EPOLLHUP errors were raised 
both at the same time.  They suggested this patch:

```
Index: twisted/internet/epollreactor.py
===================================================================
--- twisted/internet/epollreactor.py    (revision 21179)
+++ twisted/internet/epollreactor.py    (working copy)
@@ -199,6 +199,8 @@
                 if not why and event & _epoll.OUT:
                     why = selectable.doWrite()
                     inRead = False
+                if event & _epoll.HUP or event & _epoll.ERR:
+                    why = main.CONNECTION_LOST
                 if selectable.fileno() != fd:
                     why = error.ConnectionFdescWentAway(
                           'Filedescriptor went away')

```

After updating our live systems and waiting a day we now see reasonable loads.
Thanks!

Note: A similiar case seems to have been reported on the mailing list

  http://twistedmatrix.com/pipermail/twisted-python/2007-April/015211.html

Regards,
Marc

Attachments:

* [broken-twisted.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d9d/d9dbf1da845ffc9f4e9682137f334ba522fe987c/e2ad43a9ffcb3b1fd02b9ba1f7f9b6a70c54f737.py) (940 bytes) - added by exarkun on 2007-09-06 18:15:00Z - demonstration of the bug

<details><summary>Searchable metadata</summary>

```
trac-id__2809 2809
type__defect defect
reporter__mvw mvw
priority__highest highest
milestone__Core_2_6 Core-2.6
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__epoll epoll
time__1188572769000000 1188572769000000
changetime__1286996323000000 1286996323000000
version__None None
owner__ 
cc__therve cc__andi5 cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set owner to @exarkun|
|-|-|

I certainly have no idea what's going on in the guts of the epoll reactor, and it sounds like a few of our high-end users are having this problem.  So perhaps you should be the one to look at this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

HUP and ERR will be set simultaneously on a socket which has failed its connection attempt.  This doesn't seem to be the case which triggers poor behavior from epollreactor though, since IN is set as well and tcp.Client.doRead notices the connection failure and removes the socket from the reactor.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

[#2814](https://github.com/twisted/twisted/issues/2814) was closed as a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Ah, but of course, it's possible to mask off IN by registering a file descriptor so that it is only monitored for OUT events.  Then, when you would have gotten HUP|ERR|IN, you'll just get HUP|ERR.  This can happen if pauseProducing has been called on a transport.  This isn't common (possible?) for the period between when a connection is attempted and when it fails.  However, the attached demonstrates one way you can run into this for an established connection.

**Comment by twisted-trac** [comment]
|<img alt="alecm's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|alecm commented|
|-|-|

I am the person that reported this bug to the mailing list in April,

http://twistedmatrix.com/pipermail/twisted-python/2007-April/015211.html

Implementation of the patch similar to

===================================================================
--- twisted/internet/epollreactor.py    (revision 21179)
+++ twisted/internet/epollreactor.py    (working copy)
@@ -199,6 +199,8 @@
                 if not why and event & _epoll.OUT:
                     why = selectable.doWrite()
                     inRead = False
+                if event & _epoll.HUP or event & _epoll.ERR:
+                    why = main.CONNECTION_LOST

has resolved our problem, CPU is no longer 100%, epoll reactor is now usable.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

This is almost fixed in [branches/epoll-hup-err-2809](branches/epoll-hup-err-2809).  There are some tests which need to be skipped on gtkreactor though, which I haven't gotten to.

**Comment by twisted-trac** [comment]
|<img alt="alecm's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|alecm commented|
|-|-|

Replying to [exarkun](#note_9):
> This is almost fixed in [branches/epoll-hup-err-2809](branches/epoll-hup-err-2809).  There are some tests which need to be skipped on gtkreactor though, which I haven't gotten to.



We are in a production environment, so can we just substitute a single file
http://twistedmatrix.com/trac/browser/branches/epoll-hup-err-2809/twisted/internet/epollreactor.py[c]

or we need to check out this entire branch and reinstall Twisted?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

I merged the branch forward to pick up some stuff from trunk, the code is in /branches/epoll-hup-err-2809-2 now.  If you want, you can just grab http://twistedmatrix.com/trac/browser/branches/epoll-hup-err-2809-2/twisted/internet/epollreactor.py and drop it into your installation; I can't think of any reason it wouldn't work.  You might also want to get the updated test_epoll.py and test_tcp.py and run them though, just to make sure everything is working okay.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

Okay I suppose this is essentially done.  One issue I have with the code is that the new test method is quite long.  Any suggestions for shortening it (other than moving the nested classes out) would be very welcome.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set owner to @exarkun|
|-|-|

The test is a bit long, but very well commented. Once removed docs and comments, it's not that long, so I think it's OK. The fix must go in anyhow.

Please create something for the comment about gtkreactor (fwiw, I'm +1 on deprecating it), then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [21301]) Merge epoll-hup-err-2809-2

Author: exarkun
Reviewer: therve
Fixes [#2809](https://github.com/twisted/twisted/issues/2809)
Refs [[#1780](https://github.com/twisted/twisted/issues/1780)](https://github.com/twisted/twisted/issues/1780)
Refs [#11168](https://github.com/twisted/twisted/issues/11168)
Refs [[#2833](https://github.com/twisted/twisted/issues/2833)](https://github.com/twisted/twisted/issues/2833)

Change the error (EPOLLHUP, EPOLLERR) checking code so that it can
handle the case where they are both set, rather than only checking
the cases where one or the other is set.  This allows epollreactor
to properly detect and report connections which are lost uncleanly
and also are not being polled for input events.

The test case for this is slightly convoluted for a number of
reasons:

  * gtkreactor cannot pass the new test, and I am not fixing that
    reactor in this branch.  The test will skip when that reactor
    is in use.  #2833 was created as a result of this in order to
    be able to remove the broken gtkreactor eventually.
  * If a file descriptor is being polled for neither read nor write
    events, it will not be polled for hang-up events either.  This
    is a known issue, [#11168](https://github.com/twisted/twisted/issues/11168).  When it is fixed, the test will need
    to be updated.
  * In order to exercise the case where the event reported from epoll
    is only HUP|ERR (specifically, not IN), the test pauses one of
    the involved transports.  However, due to #1780, it must use a
    timing hack to accomplish this.  Once that issue is resolved,
    it should be possible to remove this hack.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
